Document renumbering lines in macros (#40955)
JuliaLang/julia: doc/src/manual/metaprogramming.md

or to insert the value as an extra argument to a runtime function call doing self-reflection
in the current module.

### Renumbering lines in macros

You will likely want to renumber new lines of code introduced in macros
Consider the following trivial macro:

```jldoctest macro_line_numbers
julia> macro adds_more_lines_1()
           quote
               "code"
           end
       end
@adds_more_lines_1 (macro with 1 method)
```

Let's take a closer look at the new code introduced by the macro:

```jldoctest macro_line_numbers
julia> numbered = quote
           "code"
       end;

julia> numbered.head
:block

julia> typeof.(numbered.args)
2-element Vector{DataType}:
 LineNumberNode
 String
```

> Julia will introduce a new line number for the new line of code pointing to the definition expression.

Reviewer: ```suggestion
When a macro is invoked, Julia will introduce a new `LineNumberNode` that points to to the definition expression inside the macro.
```

You likely want line numbering to point to the user's call site, rather than the definition expression.
This is for the following two reasons:

> - If an error occurs in code introduced in a macro, you likely want stack-traces to point to the user's call site, rather than the definition expression.

Reviewer: ```suggestion
- If an error occurs in code introduced in a macro, you likely want stack-traces to point to the user's call site, rather than the macro's internals.
```

> - If a user calls a macro, you likely want coverage to consider the user's call site as covered, rather than the definition expression.

Reviewer: ```suggestion
- If a user calls a macro, you likely want coverage to consider the user's call site as covered, rather than the macro itself.
```

Thus, consider revising the macro as follows:

```jldoctest macro_line_numbers
julia> macro adds_more_lines_2()
           result =
               quote
                   "code"
               end
           result.args[1] = __source__
           result
       end
@adds_more_lines_2 (macro with 1 method)
```

> `__source__` will contain a line number pointing to the user's call site (see more information about `__source__` above).

Reviewer: I think it would be better to link to the "macro invocation" section where the `__source__` arg is defined rather than just saying "above", since that isn't too descriptive.

> `__source__` will contain a line number pointing to the user's call site (see more information about `__source__` above).

Reviewer: ```suggestion
`__source__` will contain a `LineNumberNode` pointing to the user's call site (see more information about `__source__` above).
```

> The macro will renumber the new line of code to point the user's call site, rather than the definition expression.

Reviewer: ```suggestion
The macro will now renumber the new line of code to point the user's call site, rather than inside the macro.
```

or to insert the value as an extra argument to a runtime function call doing self-reflection
in the current module.

### Renumbering lines in macros

You will likely want to renumber new lines of code introduced in macros
Consider the following trivial macro:

```jldoctest macro_line_numbers
julia> macro adds_more_lines_1()
           quote
               "code"
           end
       end
@adds_more_lines_1 (macro with 1 method)
```

Let's take a closer look at the new code introduced by the macro:

```jldoctest macro_line_numbers
julia> numbered = quote
           "code"
       end;

julia> numbered.head
:block

julia> typeof.(numbered.args)
2-element Vector{DataType}:
 LineNumberNode
 String
```

When a macro is invoked, Julia will introduce a new `LineNumberNode` that points to inside the macro.
You likely want line numbering to point to the user's call site, rather than the definition expression.
This is for the following two reasons:

- If an error occurs in code introduced in a macro, you likely want stack-traces to point to the user's call site, rather than the macro's internals.

> - If a user calls a macro, you likely want coverage to consider the user's call site as covered, rather than the macro itself.

Reviewer 2: These argument seem inaccurate to me. The __source__ should already be be getting represented as the place where the macro is being inlined at (as a fake "caller"). Injecting it again into the macro body seems like it could lead to even stranger frames as it tries to guess why this macro is changing back and forth between inlined and local. If the source location is missing in the caller (where we splice the macro to), that seems like just a bug.

There's already `Base.remove_linenums!` and associated improvements in PR form: https://github.com/JuliaLang/julia/pull/31335 and `Meta.replace_sourceloc!` from https://github.com/JuliaLang/julia/pull/22623.

Author: I honestly understood very little of that comment...

Author: > If the source location is missing in the caller (where we splice the macro to), that seems like just a bug.

Ok, if I understand you right, a bug like this is likely what I'm running into

Author: Should I close this and open a new issue?

Thus, consider revising the macro as follows:

```jldoctest macro_line_numbers
julia> macro adds_more_lines_2()
           result =
               quote
                   "code"
               end
           result.args[1] = __source__
           result
       end
@adds_more_lines_2 (macro with 1 method)
```

> `__source__` will contain a `LineNumberNode` pointing to the user's call site.
> See [Macro invocation](@ref) for more information about `__source__`).

Reviewer: ```suggestion
`__source__` will contain a `LineNumberNode` pointing to the user's call site
(see [Macro invocation](@ref) for more information about `__source__`).
```

Minor change, since the opening parenthesis was missing.
